### PR TITLE
[None][perf] Enable in-flight batching for Nemotron3 Nano Omni multimodal encoder

### PR DESCRIPTION
**Pull request by yechank-nvidia**
## Headline numbers

| Modality | branch | main | Speedup |
|---|---:|---:|---:|
| Image (256×256, ISL=325) | **2.61 s** | 3.13 s | **1.20× (+19.9% throughput)** |
| Audio (30 s,    ISL=444) | **3.41 s** | 4.26 s | **1.25× (+25.2% throughput)** |

## Image detailed (256×256, ISL=325, CONC=32, OSL=64)

| Metric | branch run2 | branch run3 | **branch avg\*** | main run2 | main run3 | **main avg\*** | Δ (branch faster) |
|---|---:|---:|---:|---:|---:|---:|---:|
| Benchmark Duration (s)         | 2.632 | 2.597 | **2.614** | 3.134 | 3.134 | **3.134** | **-16.6%** (1.20×) |
| Request Throughput (req/s)     | 36.48 | 36.97 | **36.73** | 30.63 | 30.63 | **30.63** | **+19.9%** |
| Output Tput (tok/s)            | 2,335 | 2,366 | **2,350** | 1,961 | 1,960 | **1,961** | **+19.9%** |
| TTFT avg (ms)                  | 399.6 | 393.6 | **396.7** | 540.4 | 543.4 | **541.9** | **-26.8%** |
| TTFT p50 (ms)                  | 425.2 | 415.4 | **420.3** | 559.2 | 557.5 | **558.4** | **-24.7%** |
| TTFT p90 (ms)                  | 449.6 | 432.4 | **441.0** | 570.3 | 563.0 | **566.6** | **-22.2%** |
| TTFT p99 (ms)                  | 459.0 | 434.1 | **446.6** | 572.8 | 567.0 | **569.9** | **-21.6%** |
| Request Latency avg (ms)       | 873.6 | 862.4 | **868.0** | 1,038.7 | 1,038.5 | **1,038.6** | **-16.4%** |
| Inter-Token Latency (ms)       | 7.52  | 7.44  | **7.48**  | 7.91  | 7.86  | **7.88**  | -5.1% |

## Audio detailed (30 s @ 16 kHz, ISL=444, CONC=32, OSL=64)

| Metric | branch run1 | branch run2 | branch run3 | **branch avg** | main run1 | main run2 | main run3 | **main avg** | Δ (branch faster) |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| Benchmark Duration (s)        | 3.620   | 3.402   | 3.201   | **3.408** | 4.294   | 4.086   | 4.396   | **4.259** | **-20.0%** (1.25×) |
| Request Throughput (req/s)    | 26.52   | 28.22   | 29.99   | **28.24** | 22.35   | 23.49   | 21.84   | **22.56** | **+25.2%** |
| Output Tput (tok/s)           | 1,697   | 1,806   | 1,920   | **1,808** | 1,431   | 1,504   | 1,398   | **1,444** | **+25.2%** |
| TTFT avg (ms)                 | 636.1   | 616.3   | 477.0   | **576.5** | 827.1   | 697.2   | 812.4   | **778.9** | **-26.0%** |
| TTFT p50 (ms)                 | 655.7   | 599.3   | 477.3   | **577.4** | 872.3   | 761.4   | 930.8   | **854.8** | **-32.5%** |
| TTFT p90 (ms)                 | 813.4   | 736.9   | 645.2   | **731.8** | 1,024.3 | 883.3   | 1,022.2 | **976.6** | **-25.1%** |
| TTFT p99 (ms)                 | 815.4   | 741.7   | 647.4   | **734.8** | 1,029.0 | 884.9   | 1,037.1 | **983.7** | **-25.3%** |
| Request Latency avg (ms)      | 1,189   | 1,116   | 1,049   | **1,118** | 1,419   | 1,335   | 1,427   | **1,393** | **-19.7%** |
| Inter-Token Latency (ms)      | 8.78    | 7.93    | 9.09    | **8.60**  | 9.39    | 10.12   | 9.75    | **9.75**  | **-11.8%** |

  ## Accuracy

  | Metric | Direction | Branch (n=10) | Baseline (n=10) | Δ (branch − baseline) | Welch t | df | p (two-tailed) | Verdict |
  |---|:--:|---:|---:|---:|---:|---:|---:|---|
  | **MMMU** (image) | higher better | 38.81 ± 1.29 | 38.31 ± 1.33 | **+0.50** | 0.86 | 18 | 0.40 | no regression |
  | **VoxPopuli WER** (audio) | lower better | 6.152 ± 0.020 | 6.146 ± 0.018 | +0.006 | 0.71 | 18 | 0.49 | no regression |
  | **VideoMME** (video) | higher better | 68.43 ± 0.77 | 68.70 ± 0.79 | −0.27 | 0.77 | 18 | 0.45 | no regression |

## Release Notes

* **Performance Improvements**
  * Enhanced multimodal encoding efficiency through optimized batching of image, video, and audio inputs.
  * Improved support for advanced caching operations.

* **Tests**
  * Added end-to-end correctness tests validating batch processing for multimodal requests.
  * Expanded test coverage for audio encoding and multimodal input handling scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/13977)

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 1]
```diff
@@ -2664,52 +2758,84 @@ def _encode_multimodal(self, multimodal_params: List[MultimodalParams]) -> List[
...
                         audio_emb,
                         per_clip_audio_counts,
                         has_audio=audio_data["has_audio"],
                         audio_num_clips=audio_data["audio_num_clips"],
-                        video_sizes=video_sizes,
+                        video_sizes=param.multimodal_data["video"].get("video_size", []),
                         evs_num_tokens=(num_tokens[0] if num_tokens is not None else None),
                     )
-
                 mm_embeddings.append(vision_emb)
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Resolved.

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py [hunk 2]
```diff
@@ -691,110 +691,176 @@ def apply_evs(
...
+            pixel_values_flat = raw_pixel_values[0]
+        else:
+            pixel_values_flat = torch.cat(raw_pixel_values, dim=0)
+
+        # `extract_feature_dynamic` concatenates per-image embeddings along
+        # dim=1, so we split by cumulative per-request post-shuffle token count.
+        batched_embeds = self.extract_feature_dynamic(pixel_values_flat, image_sizes)
+        scale_squared = self.downsample_ratio**2
+        per_image_token_counts = [
+            int(round(seq_len * scale_squared))
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Nit: `int` is unnecessary here.

**Review by 2ez4bz — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_nemotron_nano.py (hunk 2 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_nemotron_nano.py]: Done.
